opm/simulators/linalg/cuistl/detail/vector_operations.cu

```diff
@@ -0,0 +1,112 @@
+/*
+  Copyright 2022-2023 SINTEF AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <opm/simulators/linalg/cuistl/detail/vector_operations.hpp>
+// TODO: [perf] Get rid of thrust.
+#include <thrust/device_ptr.h>
+#include <thrust/reduce.h>
+namespace Opm::cuistl::detail
+{
+
+namespace
+{
+    template <class T>
+    __global__ void setZeroAtIndexSetKernel(T* devicePointer, size_t numberOfElements, const int* indices)
+    {
+        const auto globalIndex = blockDim.x * blockIdx.x + threadIdx.x;
+
+        if (globalIndex < numberOfElements) {
+            devicePointer[indices[globalIndex]] = T(0);
+        }
+    }
+
+    template <class T>
+    __global__ void setVectorValueKernel(T* devicePointer, size_t numberOfElements, T value)
+    {
+        const auto globalIndex = blockDim.x * blockIdx.x + threadIdx.x;
+
+        if (globalIndex < numberOfElements) {
+            devicePointer[globalIndex] = value;
+        }
+    }
+
+    template <class T>
+    __global__ void
+    elementWiseMultiplyKernel(const T* a, const T* b, T* buffer, size_t numberOfElements, const int* indices)
+    {
+        const auto globalIndex = blockDim.x * blockIdx.x + threadIdx.x;
+
+        // TODO: [perf] Is it faster to just use a mask? Probably does not matter either way
+        //              This is hopefully not where we will spend most of our time.
+        if (globalIndex < numberOfElements) {
+            buffer[globalIndex] = a[indices[globalIndex]] * b[indices[globalIndex]];
+        }
+    }
+
+    constexpr inline size_t getThreads([[maybe_unused]] size_t numberOfElements)
+    {
+        return 1024;
```

Comment:
I assume this is a value that should work well most of the time, but is there a way to ask the hardware to tell us the optimal size?